pkg/strategy/grid2/recover.go

```diff
@@ -0,0 +1,305 @@
+package grid2
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/c9s/bbgo/pkg/bbgo"
+	"github.com/c9s/bbgo/pkg/fixedpoint"
+	"github.com/c9s/bbgo/pkg/types"
+	"github.com/pkg/errors"
+)
+
+func (s *Strategy) recoverByScanningTrades(ctx context.Context, session *bbgo.ExchangeSession) error {
+	historyService, implemented := session.Exchange.(types.ExchangeTradeHistoryService)
+	// if the exchange doesn't support ExchangeTradeHistoryService, do not run recover
+	if !implemented {
+		s.logger.Warn("ExchangeTradeHistoryService is not implemented, can not recover grid")
+		return nil
+	}
+
+	s.logger.Infof("recover grid with group id: %d", s.OrderGroupID)
+
+	openOrders, err := session.Exchange.QueryOpenOrders(ctx, s.Symbol)
+	if err != nil {
+		return errors.Wrapf(err, "check if strategy exists error when query open orders")
+	}
+
+	s.logger.Infof("found %d open orders left on the %s order book", len(openOrders), s.Symbol)
+
+	// filter out the order with the group id belongs to this grid
+	var openOrdersOnGrid []types.Order
+	for _, order := range openOrders {
+		if order.GroupID == s.OrderGroupID {
+			openOrdersOnGrid = append(openOrdersOnGrid, order)
+		}
+	}
+
+	s.logger.Infof("found %d open orders belong to this grid on the %s order book", len(openOrdersOnGrid), s.Symbol)
+
+	if s.GridProfitStats.InitialOrderID != 0 {
+		// existiting strategy, need recover
+	} else if len(openOrdersOnGrid) != 0 {
+		// open orders on grid is not 0. this strategy is existing, need recover
+	} else {
+		// initial order id may be new strategy or lost data in redis, so we need to check trades + open orders
+		// if there are open orders or trades, we need to recover
+		trades, err := historyService.QueryTrades(ctx, s.Symbol, &types.TradeQueryOptions{
+			// from 1, because some API will ignore 0 last trade id
+			LastTradeID: 1,
+			// if there is any trades, we need to recover.
+			Limit: 1,
+		})
```

Comment:
Is this safe for the new v3 API ?

Reply:
I test it on V3 API. I think it's ok.

Reply:
Can you verify this on the New v3 document? the legacy APIs are not removed yet.